### PR TITLE
Add commands to view Java Language Server debug and error logs

### LINKED ISSUE #2367: m2e logs
I have a Maven build process getting into an infinite loop state, where it's constantly building (I can see it repeatedly spawning processes that should be started once during the build). I would like to see the m2e log, like I did in Eclipse, to try to get a handle on what's going on, but I don't seem to be able to see any output from m2e. Am I missing something or is this something that we could add?

### PR DESCRIPTION
**Pull request by karlvr**
Resolves #2367, which in spite of being closed wasn't resolved as such (we found out how to view the logs but didn't make it easy)

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -720,6 +722,25 @@ function openServerLogFile(workspacePath, column: ViewColumn = ViewColumn.Active
...
 }
 
+function openRollingServerLogFile(workspacePath, filename, column: ViewColumn = ViewColumn.Active): Thenable<boolean> {
+	return new Promise((resolve) => {
+		const dirname = path.join(workspacePath, '.metadata');
+
+		// find out the newest one
+		glob(filename + '-*', { cwd: dirname }, (err, files) => {
+			if (!err && files.length > 0) {
+				files.sort();
```
[the remark this answers is not in the document]
Comment (karlvr) [reply src/extension.ts:732]: @rgrunber the file names are like `.out-jdt.ls-YYYYMMDDHHmmss.log` so I `sort`ed and ran. I guess the filenames are designed to make that the easy approach so I hope it's okay to rely upon!

File: package.nls.json [hunk 2]
```diff
@@ -3,6 +3,8 @@
 	"java.projectConfiguration.update": "Update Project",
 	"java.project.import": "Import Java Projects into Workspace",
 	"java.open.serverLog": "Open Java Language Server Log File",
+	"java.open.serverDebugLog": "Open Java Language Server Debug Log File",
+	"java.open.serverErrorLog": "Open Java Language Server Error Log File",
```
[the remark this answers is not in the document]
Comment (karlvr) [reply package.nls.json]: @rgrunber Thank you very much for this feedback. I've pushed a correction for this now.

What do you think about alerting with instructions to set `jdt.ls.debug` to `true` if those files can't be found? I'll experiment a little with that.

**Comment by testforstephen** [comment]
I'm thinking who is the audience for these logs. To me, extension developers might be interested in these logs, but normal users might not be. If this is the case, I'd rather not expose too many log commands in the Command Palette, and instead reuse the existing command `Java: Open All Log Files`. 

We can refactor this to `Java: Open All Log Files...`. And when you click on it, it will list all possible log locations for user to pick, including:
- Java Extension Client Log File
- Java Language Server Log File
- Standard Output Log File
- Standard Error Log File

@karlvr @rgrunber  WDYT?

**Comment by karlvr** [comment]
@testforstephen I started down this path because I wanted to see the logs from Maven; effectively the Maven Console I missed from Eclipse.

@rgrunber directed me to the files, including how to activate them, in #2367.

The contents of the stdout (I think) is actually what I _expected_ to see in the Maven output window in VS Code… whereas at least for me that output only ever consists of one line identifying my local repository. Given how much Maven does, seeing its output is actually something I'm regularly interested in, especially when I'm pondering how long it's taking and how much processing power it's using!

I do agree that the approach I've used in this PR is a proliferation of options.

Perhaps it would be possible to channel items from at least one of these logs into the Maven Output window, so it's directly accessible?

**Comment by rgrunber** [comment]
- I did also feel a bit like adding 2 more commands for logs might cause more confusion/clutter. So how about we remove the code that exposes the 2 additional commands, and simply have `openLogs()` call `openRollingServerLogFile(..)` twice (once for `.out-jdt.ls` & `.err-jdt.ls`). That should still improve things for now.
- What do you mean by "Maven Output window" ? I don't think we provide such a view. Is it coming from vscode-maven ? It would be different than the "Maven" output/error we produce.
- To your question of whether we could redirect the m2e stdout/stderr to somehwere else, I was thinking about this as well! At the end of the day, the only reason those 2 files exist, is because we defined them at https://github.com/eclipse/eclipse.jdt.ls/blob/master/org.eclipse.jdt.ls.core/logback.xml. But that was just the easiest thing to do at the time. There's plenty of other "Appenders" we could use. See https://github.com/eclipse-m2e/m2e-core/blob/master/org.eclipse.m2e.logback.configuration/defaultLogbackConfiguration/logback.xml as an example. The interesting one is [`EclipseLogAppender`](https://github.com/eclipse-m2e/m2e-core/blob/master/org.eclipse.m2e.logback.appender/src/org/eclipse/m2e/logback/appender/EclipseLogAppender.java). It looks like it's logging to the Eclipse Platform log, so this might actually be the way to redirect everything into the JDT-LS server logs! The slight downside.. we need to add `org.eclipse.m2e.logback.appender` (~12kb) to our JDT-LS runtime.

**Comment by karlvr** [comment]
@rgrunber sorry, too many plugins! The Maven Output is from the Maven plugin. It has an entry in the Output tab, but it's underused. It's always where I imagined the Maven log would appear. I discovered when I started my investigation there that the Maven plugin didn't control m2e but forgot!

If we can't extract just the m2e bits then getting the logs into the Java Language Server Output would be great.

**Comment by rgrunber** [comment]
Going through trying to get the EclipseLogAppender working made me discover we have an issue with custom logback configs : https://github.com/eclipse/eclipse.jdt.ls/pull/2077 , so the discussion is definitely appreciated.

I was able to get some of the events to be logged to the language server `.log` file, but not all. So until we get it working, I'm fine with accepting this. Just need to make the following change :

- Do not register the stdout/stderr commands in the command palette. They should exist so that `Open All Log Files` calls them but for now let's not expose them beyond this.

**Comment by rgrunber** [comment]
I'll be pushing this in since @karlvr did some good work to get this ready. I've used `preview: false` from [`TextDocumentOptions`](https://code.visualstudio.com/api/references/vscode-api#TextDocumentShowOptions) in order to allow the log file tabs to open under the same view column. This makes them a lot easier to read than opening each one in its own column.

The Maven related logging will show up under the server logs once the logback tracing PR is merged. It's still good to be able to view the stdout/stderr logs. Further improvements might involve adding a quick-pick menu to select which (of the 4) logs.

**Review by rgrunber — APPROVED** [review]

File: package.json [hunk 3]
```diff
@@ -924,13 +924,19 @@
       },
       {
         "command": "java.open.serverLog",
-        "title": "%java.open.serverLog%",
-        "category": "Java"
+        "when": "false"
```
Comment (testforstephen) [inline package.json:927]: Oh, you hide this command. "Open Java Language Server Log File" is the most frequently command I use for troubleshooting.

You need to update wiki as well.
https://github.com/redhat-developer/vscode-java/wiki/Troubleshooting

Comment (rgrunber) [reply package.json:927]: I"m open to restoring it, if you feel the convenience of not having the other 3 log files show up is worth it.

Comment (testforstephen) [reply package.json:927]: Yes, I'd like to keep server log command.

Comment (jdneo) [reply package.json:927]: The current change breaks the VS Code's `package.json` schema:
- title is a mandatory field here.
- `when` is not a valid field here, should be used under `menus` section.

A warning can be seen after startup:

image

Comment (testforstephen) [reply package.json:927]: Agreed. For those commands that are taken as APIs, we don't need to declare it in package.json.

**Pull request merged**